x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_action_update.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { RulesSchema } from '../../../../common/detection_engine/schemas/response/rules_schema';
+import {
+  BulkActionUpdate,
+  BulkActionUpdateType,
+} from '../../../../common/detection_engine/schemas/common/schemas';
+
+export const addItemsToArray = <T>(arr: T[], items: T[]): T[] =>
+  Array.from(new Set([...arr, ...items]));
+
+export const deleteItemsFromArray = <T>(arr: T[], items: T[]): T[] => {
+  const itemsSet = new Set(items);
+  return arr.filter((item) => !itemsSet.has(item));
```

Comment:
Why do you need `new Set(items)` here? Wouldn't `items.includes(item)` work?

Reply:
I did this to avoid N*M time complexity, where N - existing tags, M - tags to remove.
As with `.include` it would've do search through the whole array of items on every filter callback call.

Reply:
I don't think users will be removing thousands of rules for the effect of this optimization to be visible 🙂  But ok, I'll leave it up to you